### ISSUE TITLE
Using the node's key within a parameter

### ISSUE THREAD

**Issue by BaptisteFrancois** [pywr/pywr#726]
Hi, 
I am trying to create a Parameter whose return value will differ depending on the node that call this parameter for its 'max_flow' attribute. 

For instance, I have the nodes below: 
```json
"nodes": [
		{
			"name": "input_test",
			"type": "input",
			"max_flow": 4,
			"cost": -1
		},
		{
			"name": "node_1",
			"type": "output",
			"max_flow": "demand_node",
			"cost": -1
		},
		{
			"name": "node_2",
			"type": "output",
			"max_flow": "demand_node",
			"cost": -1
		}
"edges": [
		["input_test", "node_1"],
		["input_test", "node_2"]
	],
"parameters": {
               "demand_node: {
                        "type": "Demand_Node"
             },
```

The parameter  "Demand_Node" would be defined within a Python class.

The idea would be that a csv files given flow values for the differnet nodes (identified thanks to their name/key) would be initially read within this class. Then, the different value for the 'max_flow' node's attribute would be returned depending on the node's name (e.g., here either 'node_1' or 'node_2'.

Long story short - how can we 'read' the node's name/key within a python class?
Thanks!

**Comment by jetuk** [comment]
If this is just for reading different timeseries from a CSV file then I think the existing `DataframeParameter` can do this for you. See this test model for an example with constants, but it should also work with dataframes: https://github.com/pywr/pywr/blob/master/tests/models/simple_df.json

Here one CSV file with two columns is referenced by two different parameters. The "column" key is used to pull out one column from the dataframe. In this example "index" is used in the same way but for the rows - you would not need "index" for a timeseries.

To answer the actual question above. Currently there is no API available for a parameter to find out which node/attribute (if any) it is assigned to - it could be assigned to more than one. Generally if a component needs access to a specific node in its calculations we have made it take the node's name as input and it loads that as a reference to that node in the instance. This is more flexible in that a parameter can depend on state from a node other than the one it is providing data too.

**Comment by BaptisteFrancois** [comment]
It is not for just reading time series from CSV. The fact is that 'max_flow' attributes for the different nodes follow the same policy (i.e., a customized parameter defined via a python class) but differ from each other via different coefficients that are applied through the computation (that is done using a customized python class). I would like the python class to be able to access the name of the node so that it can use it to read the correct coefficient to apply through the calculation.

I understand that the python class could use the node name as input. Could be write an example below?
Thanks!

**Comment by jetuk** [comment]
Sure. There are a few ways to do this.

If I understand correctly you have a custom parameter that has some coefficients. You want to reuse the parameter class in multiple places (max_flow attributes on different nodes) in your model. However, it needs to select the appropriate coefficients based on the node's name. 

I think I would set it up something like this where the "column" key is used to filter data from "my_data.csv" - I've assumed the data is in columns here. In this way the "column" does not have to match a node name (but it could). It is up to you then how you use "my_param1", "my_param2" and "my_param3" in your nodes.

Also in this example you can specify the data  directly in JSON if you prefer. However, I don't know if you have 3 or 300 coefficients. So this might be impractical. 

Warning - I didn't really test this!

```json
    "parameters": {
        "my_param1": {
            "type": "myparameter",
            "url": "my_data.csv",
            "column": "column1"
        },
        "my_param2": {
            "type": "myparameter",
            "url": "my_data.csv",
            "column": "column2"
        },
        "my_param3": {
            "type": "myparameter",
            "c1": 1.23,
            "c2": 4.56,
            "c3": 7.89
        }        
    },
```

```python

class MyParameter(Parameter):
    def __init__(self, model, c1, c2, c3, **kwargs):
        super().__init__(model, **kwargs)
        self.c1 = c1
        self.c2 = c2
        self.c3 = c3

    def value(self, ts, si):
        return self.c1 + self.c2 ** 2 + self.c3 ** 3

    @classmethod
    def load(cls, model, data):
        # Here data contains what is in the JSON definition for each definition.
        # We use it to create an instance of the class.
        url = data.pop('url', None)

        if url is not None:
            # Here do whatever logic to read data from the url 
            # with the column identifier to filter.
            data = pandas.read_csv(url)
            column = data.pop('column')
            c1 = data.loc['c1', column]
            c2 = data.loc['c2', column]
            c3 = data.loc['c3', column]
        else:
            c1 = data.pop('c1')
            c2 = data.pop('c2')
            c3 = data.pop('c3')

        return cls(model, c1, c2, c3, **data)
MyParameter.register()

```

**Comment by rheinheimer** [comment]
I have the same issue. My reading of this is that there is no option but to explicitly create a new parameter for each node, even if the parameter / custom parameter is exactly the same, with just a different name. In my case:
```python
        "Angels PH/Water Demand/1": {
            "type": "network_PH_Water_Demand"
        },
        "Angels PH/Water Demand/2": {
            "type": "network_PH_Water_Demand"
        },
```
where these are piecewise `max_flow` for a hydropower plant. The custom parameter is the same, but the result depends on both the name of the plant, used to look up the plant capacity, and the piece number, which is used to look up the piecewise demand block size (from a table). I currently parse the parameter name in the parameter's setup function (i.e., `pp_name, attr_name, block = self.name.split('/')`, but this could also be:
```python
        "Angels PH/Water Demand/1": {
            "type": "network_PH_Water_Demand",
            "node_name": "Angels PH",
            "block": 1,
        },
        "Angels PH/Water Demand/2": {
            "type": "network_PH_Water_Demand",
            "node_name": "Angels PH",
            "block": 2,
        },
```
Still, this means I have N powerhouses x M blocks = NxN repetitions of this. So having this be reduced to a single parameter in the json file with the parameter having access to the calling node's name and fully qualified name (i.e. name and sublink, which has the sublink number in it) could be quite useful for both readability and writeability.

**Comment by jetuk** [comment]
I think this is a fundamental limitation of the parameters. They pre-compute a single value each time-step which is given to the node(s) that reference that parameter. They also, in general, do not know anything about the nodes or whatever might be using their values. 

One approach could be to have a single shared parameter that does most of the work. Then a simple reference / wrapper parameter that looks up a value from the shared parameter, but this will not really reduce the number of parameters. However, it will offload anything computationally demanding that might otherwise be repeated.

Or otherwise you use some sort of templating system (e.g. jinja2) to create your JSON file.

**Comment by snorfalorpagus** [comment]
I agree with @jetuk on this issue. Parameters have a single value that is pre-computed at the start of each timestep, that does not depend on which node (if any) makes use of them. There are some performance benefits to this as it means the value of parameters can easily be cached. It also means that value of a parameter can be recorded; what value would be recorded if more than one node accessed a parameter with a value that depended on the caller? Worse yet, what would the value of a parameter be that depended on another parameter (e.g. `ParameterThreshold`) where the other parameter value depended on the caller.

Where the number of parameters becoming difficult to manage they can be generated programatically, either via the Python API or templating JSON/YAML.
